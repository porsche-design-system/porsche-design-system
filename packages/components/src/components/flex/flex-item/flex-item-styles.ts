import type {
  FlexItemAlignSelf,
  FlexItemAlignSelfType,
  FlexItemFlex,
  FlexItemFlexType,
  FlexItemGrow,
  FlexItemGrowType,
  FlexItemOffset,
  FlexItemOffsetType,
  FlexItemShrink,
  FlexItemShrinkType,
  FlexItemWidth,
  FlexItemWidthType,
} from './flex-item-utils';
import { buildResponsiveStyles, getCss, mergeDeep } from '../../../utils';
import { addImportantToEachRule } from '../../../styles';

const flexItemWidths: { [key in Exclude<FlexItemWidthType, 'auto'>]: number } & { none: number; auto: string } = {
  none: 0,
  'one-quarter': 25,
  'one-third': 33.333333,
  half: 50,
  'two-thirds': 66.666667,
  'three-quarters': 75,
  full: 100,
  auto: 'auto',
};

export const getComponentCss = (
  width: FlexItemWidth,
  offset: FlexItemOffset,
  alignSelf: FlexItemAlignSelf,
  grow: FlexItemGrow,
  shrink: FlexItemShrink,
  flex: FlexItemFlex
): string => {
  return getCss({
    '@global': {
      ':host': addImportantToEachRule({
        boxSizing: 'border-box',
        ...mergeDeep(
          buildResponsiveStyles(width, (widthResponsive: FlexItemWidthType) => ({
            width: `${flexItemWidths[widthResponsive]}%`,
          })),
          buildResponsiveStyles(offset, (offsetResponsive: FlexItemOffsetType) => ({
            marginLeft: `${flexItemWidths[offsetResponsive]}%`,
          })),
          buildResponsiveStyles(alignSelf, (alignSelfResponsive: FlexItemAlignSelfType) => ({
            alignSelf: alignSelfResponsive,
          })),
          flex !== 'initial' // flex shorthand conflicts with grow and shrink, which means even default grow or shrink props would override flex
            ? buildResponsiveStyles(flex, (flexResponsive: FlexItemFlexType) => ({
                flex: flexResponsive === 'equal' ? '1 1 0' : flexResponsive,
              }))
            : mergeDeep(
                buildResponsiveStyles(grow, (flexGrow: FlexItemGrowType) => ({ flexGrow })),
                buildResponsiveStyles(shrink, (flexShrink: FlexItemShrinkType) => ({ flexShrink }))
              )
        ),
      }),
    },
  });
};
