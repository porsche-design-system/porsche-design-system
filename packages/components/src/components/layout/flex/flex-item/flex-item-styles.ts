import type {
  FlexItemAlignSelf,
  FlexItemAlignSelfType,
  FlexItemFlex,
  FlexItemFlexType,
  FlexItemGrow,
  FlexItemGrowType,
  FlexItemOffset,
  FlexItemOffsetType,
  FlexItemShrink,
  FlexItemShrinkType,
  FlexItemWidth,
  FlexItemWidthType,
} from './flex-item-utils';
import type { GetStylesFunction, JssStyle } from '../../../../utils';
import { buildResponsiveHostStyles, getCss, mergeDeep } from '../../../../utils';
import { addImportantToEachRule } from '../../../../styles/styles';

const flexItemWidths: { [key in Exclude<FlexItemWidthType, 'auto'>]: number } & { none: number; auto: string } = {
  none: 0,
  'one-quarter': 25,
  'one-third': 33.333333,
  half: 50,
  'two-thirds': 66.666667,
  'three-quarters': 75,
  full: 100,
  auto: 'auto',
};

const getWidthStyles: GetStylesFunction = (width: FlexItemWidthType): JssStyle => ({
  width: `${flexItemWidths[width]}%`,
});

const getOffsetStyles: GetStylesFunction = (offset: FlexItemOffsetType): JssStyle => ({
  marginLeft: `${flexItemWidths[offset]}%`,
});

const getAlignSelfStyles: GetStylesFunction = (alignSelf: FlexItemAlignSelfType): JssStyle => ({ alignSelf });

const getGrowStyles: GetStylesFunction = (flexGrow: FlexItemGrowType): JssStyle => ({ flexGrow });

const getShrinkStyles: GetStylesFunction = (flexShrink: FlexItemShrinkType): JssStyle => ({ flexShrink });

const getFlexStyles: GetStylesFunction = (flex: FlexItemFlexType): JssStyle => ({
  flex: flex === 'equal' ? '1 1 0' : flex,
});

export const getComponentCss = (
  width: FlexItemWidth,
  offset: FlexItemOffset,
  alignSelf: FlexItemAlignSelf,
  grow: FlexItemGrow,
  shrink: FlexItemShrink,
  flex: FlexItemFlex
): string => {
  return getCss(
    addImportantToEachRule(
      mergeDeep(
        {
          ':host': {
            boxSizing: 'border-box',
          },
        },
        buildResponsiveHostStyles(width, getWidthStyles),
        buildResponsiveHostStyles(offset, getOffsetStyles),
        buildResponsiveHostStyles(alignSelf, getAlignSelfStyles),
        flex !== 'initial' // flex shorthand conflicts with grow and shrink, which means even default grow or shrink props would override flex
          ? buildResponsiveHostStyles(flex, getFlexStyles)
          : mergeDeep(
              buildResponsiveHostStyles(grow, getGrowStyles),
              buildResponsiveHostStyles(shrink, getShrinkStyles)
            )
      )
    )
  );
};
