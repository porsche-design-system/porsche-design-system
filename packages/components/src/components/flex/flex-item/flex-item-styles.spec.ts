import { getComponentCss } from './flex-item-styles';
import type {
  FlexItemAlignSelf,
  FlexItemFlex,
  FlexItemGrow,
  FlexItemOffset,
  FlexItemShrink,
  FlexItemWidth,
} from './flex-item-utils';
import {
  FLEX_ITEM_ALIGN_SELFS,
  FLEX_ITEM_FLEXS,
  FLEX_ITEM_GROWS,
  FLEX_ITEM_OFFSETS,
  FLEX_ITEM_SHRINKS,
  FLEX_ITEM_WIDTHS,
} from './flex-item-utils';

describe('getComponentCss()', () => {
  const dataWidth: FlexItemWidth[] = [
    ...FLEX_ITEM_WIDTHS,
    { base: 'auto', xs: 'half', s: 'auto', m: 'half', l: 'auto', xl: 'half' },
  ];
  it.each<FlexItemWidth>(dataWidth)('should return correct css for width: %j', (width) => {
    expect(getComponentCss(width, 'none', 'auto', 0, 1, 'initial')).toMatchSnapshot();
  });

  const dataOffset: FlexItemOffset[] = [
    ...FLEX_ITEM_OFFSETS,
    { base: 'one-third', xs: 'half', s: 'one-third', m: 'half', l: 'one-third', xl: 'half' },
  ];
  it.each<FlexItemOffset>(dataOffset)('should return correct css for offset: %j', (offset) => {
    expect(getComponentCss('auto', offset, 'auto', 0, 1, 'initial')).toMatchSnapshot();
  });

  const dataAlignSelf: FlexItemAlignSelf[] = [
    ...FLEX_ITEM_ALIGN_SELFS,
    { base: 'stretch', xs: 'center', s: 'stretch', m: 'center', l: 'stretch', xl: 'center' },
  ];
  it.each<FlexItemAlignSelf>(dataAlignSelf)('should return correct css for alignSelf: %j', (alignSelf) => {
    expect(getComponentCss('auto', 'none', alignSelf, 0, 1, 'initial')).toMatchSnapshot();
  });

  const dataGrow: FlexItemGrow[] = [...FLEX_ITEM_GROWS, { base: 1, xs: 0, s: 1, m: 0, l: 1, xl: 0 }];
  it.each<FlexItemGrow>(dataGrow)('should return correct css for grow: %j', (grow) => {
    expect(getComponentCss('auto', 'none', 'auto', grow, 1, 'initial')).toMatchSnapshot();
  });

  const dataShrink: FlexItemShrink[] = [...FLEX_ITEM_SHRINKS, { base: 0, xs: 1, s: 0, m: 1, l: 0, xl: 1 }];
  it.each<FlexItemShrink>(dataShrink)('should return correct css for shrink: %j', (shrink) => {
    expect(getComponentCss('auto', 'none', 'auto', 0, shrink, 'initial')).toMatchSnapshot();
  });

  const dataFlex: FlexItemFlex[] = [
    ...FLEX_ITEM_FLEXS,
    { base: 'auto', xs: 'equal', s: 'auto', m: 'equal', l: 'auto', xl: 'equal' },
  ];
  it.each<FlexItemFlex>(dataFlex)('should return correct css for flex: %j', (flex) => {
    expect(getComponentCss('auto', 'none', 'auto', 0, 1, flex)).toMatchSnapshot();
  });

  it('should return correct css for all props being breakpoint customizable', () => {
    expect(
      getComponentCss(
        [...dataWidth].pop(),
        [...dataOffset].pop(),
        [...dataAlignSelf].pop(),
        [...dataGrow].pop(),
        [...dataShrink].pop(),
        [...dataFlex].pop()
      )
    ).toMatchSnapshot();
  });
});
