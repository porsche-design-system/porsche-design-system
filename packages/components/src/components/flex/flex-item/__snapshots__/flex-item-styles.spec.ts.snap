// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`getComponentCss() should return correct css for alignSelf: "auto" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for alignSelf: "baseline" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: baseline !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for alignSelf: "center" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: center !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for alignSelf: "flex-end" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: flex-end !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for alignSelf: "flex-start" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: flex-start !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for alignSelf: "stretch" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: stretch !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for alignSelf: {"base":"stretch","xs":"center","s":"stretch","m":"center","l":"stretch","xl":"center"} 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: stretch !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}
@media(min-width:480px) {
  :host {
    align-self: center !important;
  }
}

@media(min-width:760px) {
  :host {
    align-self: stretch !important;
  }
}

@media(min-width:1000px) {
  :host {
    align-self: center !important;
  }
}

@media(min-width:1300px) {
  :host {
    align-self: stretch !important;
  }
}

@media(min-width:1760px) {
  :host {
    align-self: center !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for all props being breakpoint customizable 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 33.333333% !important;
  align-self: stretch !important;
  flex: auto !important;
}
@media(min-width:480px) {
  :host {
    width: 50% !important;
    margin-left: 50% !important;
    align-self: center !important;
    flex: 1 1 0 !important;
  }
}

@media(min-width:760px) {
  :host {
    width: auto% !important;
    margin-left: 33.333333% !important;
    align-self: stretch !important;
    flex: auto !important;
  }
}

@media(min-width:1000px) {
  :host {
    width: 50% !important;
    margin-left: 50% !important;
    align-self: center !important;
    flex: 1 1 0 !important;
  }
}

@media(min-width:1300px) {
  :host {
    width: auto% !important;
    margin-left: 33.333333% !important;
    align-self: stretch !important;
    flex: auto !important;
  }
}

@media(min-width:1760px) {
  :host {
    width: 50% !important;
    margin-left: 50% !important;
    align-self: center !important;
    flex: 1 1 0 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for flex: "auto" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex: auto !important;
}"
`;

exports[`getComponentCss() should return correct css for flex: "equal" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex: 1 1 0 !important;
}"
`;

exports[`getComponentCss() should return correct css for flex: "initial" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for flex: "none" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex: none !important;
}"
`;

exports[`getComponentCss() should return correct css for flex: {"base":"auto","xs":"equal","s":"auto","m":"equal","l":"auto","xl":"equal"} 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex: auto !important;
}
@media(min-width:480px) {
  :host {
    flex: 1 1 0 !important;
  }
}

@media(min-width:760px) {
  :host {
    flex: auto !important;
  }
}

@media(min-width:1000px) {
  :host {
    flex: 1 1 0 !important;
  }
}

@media(min-width:1300px) {
  :host {
    flex: auto !important;
  }
}

@media(min-width:1760px) {
  :host {
    flex: 1 1 0 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for grow: {"base":1,"xs":0,"s":1,"m":0,"l":1,"xl":0} 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 1 !important;
  flex-shrink: 1 !important;
}
@media(min-width:480px) {
  :host {
    flex-grow: 0 !important;
  }
}

@media(min-width:760px) {
  :host {
    flex-grow: 1 !important;
  }
}

@media(min-width:1000px) {
  :host {
    flex-grow: 0 !important;
  }
}

@media(min-width:1300px) {
  :host {
    flex-grow: 1 !important;
  }
}

@media(min-width:1760px) {
  :host {
    flex-grow: 0 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for grow: 0 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for grow: 1 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 1 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: "half" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 50% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: "none" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: "one-quarter" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 25% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: "one-third" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 33.333333% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: "three-quarters" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 75% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: "two-thirds" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 66.666667% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for offset: {"base":"one-third","xs":"half","s":"one-third","m":"half","l":"one-third","xl":"half"} 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 33.333333% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}
@media(min-width:480px) {
  :host {
    margin-left: 50% !important;
  }
}

@media(min-width:760px) {
  :host {
    margin-left: 33.333333% !important;
  }
}

@media(min-width:1000px) {
  :host {
    margin-left: 50% !important;
  }
}

@media(min-width:1300px) {
  :host {
    margin-left: 33.333333% !important;
  }
}

@media(min-width:1760px) {
  :host {
    margin-left: 50% !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for shrink: {"base":0,"xs":1,"s":0,"m":1,"l":0,"xl":1} 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 0 !important;
}
@media(min-width:480px) {
  :host {
    flex-shrink: 1 !important;
  }
}

@media(min-width:760px) {
  :host {
    flex-shrink: 0 !important;
  }
}

@media(min-width:1000px) {
  :host {
    flex-shrink: 1 !important;
  }
}

@media(min-width:1300px) {
  :host {
    flex-shrink: 0 !important;
  }
}

@media(min-width:1760px) {
  :host {
    flex-shrink: 1 !important;
  }
}
"
`;

exports[`getComponentCss() should return correct css for shrink: 0 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 0 !important;
}"
`;

exports[`getComponentCss() should return correct css for shrink: 1 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "auto" 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "full" 1`] = `
":host {
  box-sizing: border-box !important;
  width: 100% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "half" 1`] = `
":host {
  box-sizing: border-box !important;
  width: 50% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "one-quarter" 1`] = `
":host {
  box-sizing: border-box !important;
  width: 25% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "one-third" 1`] = `
":host {
  box-sizing: border-box !important;
  width: 33.333333% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "three-quarters" 1`] = `
":host {
  box-sizing: border-box !important;
  width: 75% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: "two-thirds" 1`] = `
":host {
  box-sizing: border-box !important;
  width: 66.666667% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}"
`;

exports[`getComponentCss() should return correct css for width: {"base":"auto","xs":"half","s":"auto","m":"half","l":"auto","xl":"half"} 1`] = `
":host {
  box-sizing: border-box !important;
  width: auto% !important;
  margin-left: 0% !important;
  align-self: auto !important;
  flex-grow: 0 !important;
  flex-shrink: 1 !important;
}
@media(min-width:480px) {
  :host {
    width: 50% !important;
  }
}

@media(min-width:760px) {
  :host {
    width: auto% !important;
  }
}

@media(min-width:1000px) {
  :host {
    width: 50% !important;
  }
}

@media(min-width:1300px) {
  :host {
    width: auto% !important;
  }
}

@media(min-width:1760px) {
  :host {
    width: 50% !important;
  }
}
"
`;
